### DISCUSSION TITLE
Slow computation when when using tensor mechanics

### DISCUSSION THREAD

**Discussion by geoyanzhan3** [idaholab/moose#20513] [Q&A General] [answered]
Hi, 

I am benchmarking MOOSE with COMSOL Multiphysics. 

When solving the same problem, MOOSE takes 10x more time than COMSOL. Both of them used 8 cores.

| Software  |  Num Elems   |   DOF     |   Time |
|:---------:|:---------:|:---------:|:---------:|
| MOOSE     |     14152         |       63714  |     40+s |
| COMSOL   |    22390       |        98910   |     4s | 

I wonder if I miss something or my input setting needs optimization.

I have attached the input files and meshes.

Thank you very much!

Yan

[TensorMechanics_speedup.zip](https://github.com/idaholab/moose/files/8216043/TensorMechanics_speedup.zip)

```
[Mesh]
  file = box3d_Grav0DefM1SOrd1P2_Nelem1.msh
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = SMALL
    add_variables = true
  []
[]

[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 30e9
    poissons_ratio = 0.25
  []
  [elastic_stress]
    type = ComputeLinearElasticStress
  []
[]

[BCs]
  [bottom]
    type = DirichletBC
    variable = disp_z
    boundary = bottom
    value = 0
  []
  [lf]
    type = DirichletBC
    variable = disp_x
    boundary = leftright
    value = 0
  []
  [fb]
    type = DirichletBC
    variable = disp_y
    boundary = frontback
    value = 0
  []
  [pressure_x]
    type = Pressure
    boundary = chamber
    variable = disp_x
    component = 0
    factor = 1e6
    use_displaced_mesh = true
  []
  [pressure_y]
    type = Pressure
    boundary = chamber
    variable = disp_y
    component = 1
    factor = 1e6
    use_displaced_mesh = true
  []
  [pressure_z]
    type = Pressure
    boundary = chamber
    variable = disp_z
    component = 2
    factor = 1e6
    use_displaced_mesh = true
  []
[]


[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Steady

  solve_type = 'NEWTON'

  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-pc_type -sub_pc_type -pc_asm_overlap -ksp_gmres_restart'
  petsc_options_value = 'asm lu 1 101'
[]


[Outputs]
  exodus = true
  perf_graph = true
  csv = false
[]

```

**Comment by jiangwen84** [comment, marked as answer]
@geoyanzhan3 The number of elements and dofs are different.  Any reason why you use different mesh or element types to solve the same problem? For this small problem, using 8 cores is not optimal. What linear solver does Comsol use?

**Reply by geoyanzhan3** [reply]
@jiangwen84 Due to COMSOL does not support using the meshes created by gmsh. Also, the element types of COMSOL and MOOSE should be the same (tet 10 nodes). 

And if I don't use "mpiexec -n 8", it took MOOSE for >250s. COMSOL is using 8 cores. So I use 8 cores for MOOSE as well.

It seems [COMSOL](https://www.comsol.com/blogs/solutions-linear-systems-equations-direct-iterative-solvers/) is using PARDISO. I have also tested MUMP. It took 4s to finish as well.

**Reply by GiudGiud** [reply]
@geoyanzhan3 you need to use the same linear solver in both applications.
and the same mesh. The size and shape of the cells influence the difficulty of solving the linear system.

As @jiangwen84 mentioned 8 cores is not optimal, it brings the simulation to ~8k dofs per processor. Our guideline is at least 20k

also, are you using the `conda` installation? It's not optimized for your machine, it's not architecture specific. You ll need to install your own MPI, petsc and libmesh to get better performance on your machine.

**Reply by geoyanzhan3** [reply]
@GiudGiud @jiangwen84 Do you know if MOOSE supports any of these mesh files from COMSOL?

<img width="469" alt="Screen Shot 2022-03-09 at 11 45 16 AM" src="https://user-images.githubusercontent.com/35808906/157489433-73046c5e-3e06-4c60-88ae-54f55019e32b.png">

**Reply by geoyanzhan3** [reply]
@GiudGiud And, I followed this link https://mooseframework.inl.gov/getting_started/installation/conda.html, when installing the MOOSE. Is that what you do not recommend?

**Reply by GiudGiud** [reply]
does not look like it
Here's the list of mesh formats we support
https://mooseframework.inl.gov/moose/source/mesh/FileMesh.html

That installation method is fine (great) for regular use. For benchmarking, I d follow that one instead:
https://mooseframework.inl.gov/getting_started/installation/manual_installation_gcc.html

**Reply by jiangwen84** [reply]
1.  You can try MUMPS solver with MOOSE. I forgot if MUMPS is installed by default or not though. 
```
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = ' lu       mumps'
```
2. I do not see any COMSOL supported mesh files can be directly imported to MOOSE. Maybe we can try a simple problem  (e.g., cube) that can be meshed in a same way in MOOSE and COMSOL. 
3.  Take a closer look at the `perf_graph`. There is computational cost in problem setup, output, besides solving the system. So we want to make sure we are doing apples-to-apples comparison.

**Reply by geoyanzhan3** [reply]
@jiangwen84 Thank you!

I do see the solving time reduce from 40s to 7s when using:

```
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = ' lu       mumps'
```
Unfortunately, there is no obvious speed up when using:
```
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
```
as @hugary1995 suggested. Make me wonder if I should manually install gcc/mpich? Can it be the reason @GiudGiud ?

**Reply by jiangwen84** [reply]
If you do not specify the `pc_factor_mat_solver_package` as @hugary1995 suggested, I believe the default will be `superlu_dist` for parallel runs. I have not seen any major difference in performance between `MUMPS` and `superlu_dist`. Maybe something is wrong with your installation as @GiudGiud mentioned.

**Reply by GiudGiud** [reply]
nothing is wrong afaik. It's just not as fast to install this way. 

He's running serial in this case. So not sure it's that default

**Reply by geoyanzhan3** [reply]
I think @jiangwen84 and @hugary1995 's suggestion has already helped a lot! I appreciate that! Also many thanks to @GiudGiud! I will try to use the mesh to conduct the benchmark. Please let me know if there is a good way to make COMSOL and MOOSE use the same mesh. This work aims to introduce MOOSE to earth science (volcanology) and make FEM more accessible to people who cannot afford expensive, commercial software

**Reply by hugary1995** [reply]
Based on my own experience several years ago, MOOSE scales better than COMSOL. Although I am sure COMSOL has improved since then.

Back then, COMSOL scales reasonably well up to 8 cores, but it has diminishing speedups with more cores (Don't quote me on this. Again, I tested this several years ago, and the result is apparently model dependent). While MOOSE scales very well up to thousands of cores.

**Reply by geoyanzhan3** [reply]
I just used Gmsh to generate the same mesh for both MOOSE and COMSOL. For COMSOL, you need to save the mesh file as NASTRAN BDF file. 

The solution times for COMSOL and MOOSE are identical with using 8 cores. The number of DOF is 63714. 

|Tests | time (1 cpu)| time (8 cpu)|
|:--:|:--:|:--:|
|MOOSE | 15.88  s | 4.97 s |
|COMSOL | ~7 s | ~5 s |

It seems COMSOL may not have a better performance with more cores.

**Reply by jiangwen84** [reply]
It looks much better now, but I am still curious to figure out why MOOSE takes more than two times CPU time when using a  single core. Do you have any detail information of the performance? e.g., number of nonlinear iteration, solving time, assembly time, etc.

**Reply by geoyanzhan3** [reply]
@jiangwen84 

Here is what I get from MOOSE:

```
Setting Up
  Finished Initializing                                                                  [  0.74 s] [  104 MB]
Finished Setting Up                                                                      [  0.96 s] [  130 MB]
Framework Information:
MOOSE Version:           git commit 0c14bf0c80 on 2021-09-09
LibMesh Version:         
PETSc Version:           3.15.1
SLEPc Version:           3.15.1
Current Time:            Wed Mar  9 17:20:07 2022
Executable Timestamp:    Wed Sep 22 11:05:21 2021

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   21238
  Elems:                   14152
  Num Subdomains:          1

Nonlinear System:
  Num DOFs:                63714
  Num Local DOFs:          63714
  Variables:               { "disp_x" "disp_y" "disp_z" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "SECOND" 

Execution Information:
  Executioner:             Steady
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    SMP

 0 Nonlinear |R| = 2.841364e+11
      0 Linear |R| = 2.841364e+11
      1 Linear |R| = 5.310088e-03
 1 Nonlinear |R| = 1.556627e+07
      0 Linear |R| = 1.556627e+07
      1 Linear |R| = 2.391911e-07
 2 Nonlinear |R| = 6.478534e+02
 Solve Converged!
Executing
  Finished Solving                                                                       [ 15.47 s] [  265 MB]

Outlier Variable Residual Norms:
  disp_z: 5.570158e+02


Performance Graph:
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|                                Section                               | Calls |   Self(s)  |   Avg(s)   |    %   | Mem(MB) |  Total(s)  |   Avg(s)   |    %   | Mem(MB) |
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------
| TensorMechanicsTestApp (main)                                        |     1 |      0.009 |      0.009 |   0.05 |       1 |     16.852 |     16.852 | 100.00 |     410 |
|   Action::SetupMeshAction::Mesh::SetupMeshAction::act::setup_mesh    |     1 |      0.000 |      0.000 |   0.00 |       0 |      0.000 |      0.000 |   0.00 |       0 |
|   Action::SetupMeshAction::Mesh::SetupMeshAction::act::set_mesh_base |     2 |      0.020 |      0.010 |   0.12 |       6 |      0.127 |      0.064 |   0.75 |      14 |
|   FEProblem::outputStep                                              |     2 |      0.000 |      0.000 |   0.00 |       0 |      0.137 |      0.068 |   0.81 |       6 |
|     Exodus::outputStep                                               |     2 |      0.085 |      0.042 |   0.50 |       6 |      0.085 |      0.042 |   0.50 |       6 |
|   Steady::PicardSolve                                                |     1 |      0.000 |      0.000 |   0.00 |       0 |     15.471 |     15.471 |  91.80 |     265 |
|     FEProblem::outputStep                                            |     1 |      0.000 |      0.000 |   0.00 |       0 |      0.000 |      0.000 |   0.00 |       0 |
|     FEProblem::solve                                                 |     1 |      7.028 |      7.028 |  41.71 |     208 |     15.470 |     15.470 |  91.80 |     265 |
|       FEProblem::computeResidualInternal                             |     1 |      0.000 |      0.000 |   0.00 |       0 |      0.204 |      0.204 |   1.21 |       1 |
|       FEProblem::computeResidualInternal                             |     3 |      0.000 |      0.000 |   0.00 |       0 |      0.645 |      0.215 |   3.83 |       7 |
|       FEProblem::computeJacobianInternal                             |     2 |      0.000 |      0.000 |   0.00 |       0 |      7.591 |      3.795 |  45.04 |      49 |
|   Steady::final                                                      |     1 |      0.000 |      0.000 |   0.00 |       0 |      0.000 |      0.000 |   0.00 |       0 |
|     FEProblem::outputStep                                            |     1 |      0.000 |      0.000 |   0.00 |       0 |      0.000 |      0.000 |   0.00 |       0 |
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Finished Executing                                                                       [ 15.88 s] [  279 MB]

```

**Reply by geoyanzhan3** [reply]
Not much information from the COMSOL:

```
================================================================================
===================== Opened Elast_benchmark_3D_import.mph =====================
================================================================================
<---- Compile Equations: Stationary in Study 1/Solution 1 (sol1) ---------------
Started at Mar 9, 2022, 5:41:28 PM.
Geometry shape function: Quadratic serendipity
Running on Intel(R) Core(TM) i7-10700K CPU at 3.80 GHz.
Using 1 socket with 1 core in total on Yan-Zhan-iMac.local.
Available memory: 32.77 GB.
Time: 1 s.
Physical memory: 1.42 GB
Virtual memory: 11.55 GB
Ended at Mar 9, 2022, 5:41:28 PM.
----- Compile Equations: Stationary in Study 1/Solution 1 (sol1) -------------->
<---- Dependent Variables 1 in Study 1/Solution 1 (sol1) -----------------------
Started at Mar 9, 2022, 5:41:28 PM.
Solution time: 0 s.
Physical memory: 1.44 GB
Virtual memory: 11.57 GB
Ended at Mar 9, 2022, 5:41:29 PM.
----- Dependent Variables 1 in Study 1/Solution 1 (sol1) ---------------------->
<---- Stationary Solver 1 in Study 1/Solution 1 (sol1) -------------------------
Started at Mar 9, 2022, 5:41:29 PM.
Linear solver
Number of degrees of freedom solved for: 63714.
Symmetric matrices found.
Scales for dependent variables:
Displacement field (comp1.u): 1
Orthonormal null-space function used.
Iter      SolEst     Damping    Stepsize #Res #Jac #Sol   LinErr   LinRes
   1        0.74   1.0000000        0.74    1    1    1  5.8e-15  1.3e-14
Solution time: 5 s.
Physical memory: 2.23 GB
Virtual memory: 12.4 GB
Ended at Mar 9, 2022, 5:41:33 PM.
----- Stationary Solver 1 in Study 1/Solution 1 (sol1) ------------------------>
================================================================================
```

**Reply by hugary1995** [reply]
Well, the matrices in MOOSE and COMSOL should be very similar despite different dof maps. I think this suggests that PARDISO is faster than PETSc with the options we are using.

**Reply by jiangwen84** [reply]
@hugary1995  are you familiar with COMSOL output? Does the output tell that only 1 iteration is performed? It seems MOOSE takes two iterations, and that might explain the 2x difference there.

**Reply by hugary1995** [reply]
Yeah this is one iteration in COMSOL. But in MOOSE we should also see one iteration. I wonder if automatic scaling would help here. Also you could try applying the pressures on the undisplaced mesh.

**Reply by YaqiWang** [reply]
One iteration may not be enough because the residual norm only drops by a factor of 10^4.

**Comment by hugary1995** [comment]
Using the following options bring the solve time from 23.7 seconds down to 3.1 seconds on my laptop.
```python
[Executioner]
  type = Steady

  solve_type = 'NEWTON'

  # petsc_options = '-snes_ksp_ew'
  # petsc_options_iname = '-pc_type -sub_pc_type -pc_asm_overlap -ksp_gmres_restart'
  # petsc_options_value = 'asm lu 1 101'

  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
[]
```

We should really put a big bold emphasis somewhere in the tensor mechanics module doc saying that "using a direct solver, e.g. superlu_dist, mumps, strumpack, etc. is strooooongly recommended".

**Reply by GiudGiud** [reply]
Depends on your problem size I reckon. This is a really small problem, so LU scaling is not an issue.

**Reply by hugary1995** [reply]
Yeah, let's put a conditional so it reads:
> "Using a direct solver, e.g. superlu_dist, mumps, strumpack, etc. is strooooongly recommended if memory isn't a concern".

**Reply by GiudGiud** [reply]
3.1s though. Are we faster than COMSOL now @geoyanzhan3  ?

**Reply by hugary1995** [reply]
With his `box3d_Grav0DefM1SOrd1P2_Nelem1.msh` mesh with 34791 dofs, the solve time reduces from 24s to 3.1s.
With his `box3d_Grav0DefM1SOrd1P2_Nelem4.msh` mesh with 63714 dofs, the solve time reduces from 56.8s to 9.5s.

**Reply by geoyanzhan3** [reply]
@hugary1995 Thank you! Have you changed other settings than 'petsc_options'? Why was my solve time only reduced from 40.27s to 40.03? I am using a single core.

Here is the new code: 
```
[Mesh]
  file = box3d_Grav0DefM1SOrd1P2_Nelem1.msh
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = SMALL
    add_variables = true
  []
[]

[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 30e9
    poissons_ratio = 0.25
  []
  [elastic_stress]
    type = ComputeLinearElasticStress
  []
[]

[BCs]
  [bottom]
    type = DirichletBC
    variable = disp_z
    boundary = bottom
    value = 0
  []
  [lf]
    type = DirichletBC
    variable = disp_x
    boundary = leftright
    value = 0
  []
  [fb]
    type = DirichletBC
    variable = disp_y
    boundary = frontback
    value = 0
  []
  [pressure_x]
    type = Pressure
    boundary = chamber
    variable = disp_x
    component = 0
    factor = 1e6
    use_displaced_mesh = true
  []
  [pressure_y]
    type = Pressure
    boundary = chamber
    variable = disp_y
    component = 1
    factor = 1e6
    use_displaced_mesh = true
  []
  [pressure_z]
    type = Pressure
    boundary = chamber
    variable = disp_z
    component = 2
    factor = 1e6
    use_displaced_mesh = true
  []
[]


[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Steady

  solve_type = 'NEWTON'

  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  
[]


[Outputs]
  exodus = true
  perf_graph = true
  csv = false
[]

```

**Reply by hugary1995** [reply]
I haven't changed other settings. I was using 8 cores.

**Reply by geoyanzhan3** [reply]
Thank you @hugary1995 ! 

It works on my computer, when specify the "pc_factor_mat_solver_package".

```
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = ' lu       mumps'
```
